/* references
    - guide: https://developers.google.com/protocol-buffers/docs/proto
    - using timestamp: https://stackoverflow.com/questions/3574716/date-and-time-type-for-use-with-protobuf

*/

syntax = "proto3";
package sensors;
import "google/protobuf/timestamp.proto";

message Gps{
    double latitude = 1;
    double longitude = 2;
    double horizontalDilutionOfPrecision = 3;
    double timestamp = 4;
}

message Imu{
    float scaledAccelX = 1;
    float scaledAccelY = 2;
    float scaledAccelZ = 3;
    double timestamp = 4;
}

message Cam{
    int32 cols = 1;
    int32 rows = 2;
    bytes image_data = 3;
    double timestamp = 4;
}

message Lidar{
    float size=1;
     message xyz{
    float x=1;
    float y=2;
    float z=3;
    }
    repeated xyz data=2;
    double timestamp = 3;
}

message Can{
    float type = 1;
    float data = 2;
    double timestamp = 3;

}

message Log{
    
    uint32 can_id = 2 ;  /* 32 bit CAN_ID + EFF/RTR/ERR flags */
	uint32    can_dlc = 3; /* frame payload length in byte (0 .. CAN_MAX_DLEN) */
	bytes    __pad = 4 ;   /* padding */
	bytes    __res0 = 5 ;  /* reserved / padding */
	bytes    __res1 = 6 ;  /* reserved / padding */
	bytes    data = 7;

}

message algorithm_img
{
    int32 cols = 1;
    int32 rows = 2;
    bytes image_original = 3;
    bytes image_result = 4;
    float millis_term = 5;
}

enum ChannelOrder {
	GRAYSCALE = 0;
	BGR = 1;
	RGB = 2;
	BGRA = 3;
	RGBA = 4;
	OPTICAL_FLOW = 5;
} 